File: 프로그래머스/lv2/87946. 피로도/피로도.py
def solution(k, dungeons): 
    global answer
    answer = -1
    # 방문여부 리스트
    visited = [0] * len(dungeons)
    dfs(k,dungeons,visited,0)
    return answer

def dfs(k, dungeons,visited, cnt) :
    global answer
    
    # 방문한 던전의 수 업데이트
    if answer < cnt : answer=cnt

    for i in range(len(dungeons)) :
        # 현재 필요도로 방문 가능하고, 방문하지 않은 경우 방문
        if k >= dungeons[i][0] and not visited[i]:
            visited[i]=1
            dfs(k-dungeons[i][1],dungeons,visited,cnt+1)
            visited[i]=0

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    var answer = -1;
    const visited = Array(dungeons.length).fill(false);
    
    const dfs = (현재,방문수,visited) => {
        if(방문수>answer) answer=방문수;
        
        for(let i=0; i<dungeons.length;i++) {
            if(!visited[i] && 현재>=dungeons[i][0]) {
                visited[i]=true;
                dfs(현재-dungeons[i][1],방문수+1,visited)
                visited[i]=false;
            }
        }
    }
    
    dfs(k,0,visited)
    
    return answer;
}